File: 2326/Solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func spiralMatrix(m int, n int, head *ListNode) [][]int {
    topRow, rightCol, bottomRow, leftCol := 0, n - 1, m - 1, 0
    matrix := make([][]int, m)

    for i := range matrix {
        matrix[i] = make([]int, n)
        for j := range matrix[i] {
            matrix[i][j] = -1
        }
    }

    for head != nil {
        for i := leftCol; i <= rightCol && head != nil; i++ {
            matrix[topRow][i] = head.Val
            head = head.Next
        }
        topRow++

        for i := topRow; i <= bottomRow && head != nil; i++ {
            matrix[i][rightCol] = head.Val
            head = head.Next
        }
        rightCol--

        for i := rightCol; i >= leftCol && head != nil; i-- {
            matrix[bottomRow][i] = head.Val
            head = head.Next
        }
        bottomRow--

        for i:= bottomRow; i >= topRow && head != nil; i-- {
            matrix[i][leftCol] = head.Val
            head = head.Next
        }
        leftCol++
    }
    return matrix
    
}

File: 2326/Solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {number} m
 * @param {number} n
 * @param {ListNode} head
 * @return {number[][]}
 */
var spiralMatrix = function(m, n, head) {
    const matrix = Array.from({ length: m }, () => Array(n).fill(-1));

    let topRow = 0, bottomRow = m -1, leftCol = 0, rightCol = n - 1;

    while(head){
        for(let col = leftCol; col <= rightCol && head; col++){
            matrix[topRow][col] = head.val;
            head= head.next;
        }
        topRow++;

        for(let row = topRow; row <= bottomRow && head; row++){
            matrix[row][rightCol] = head.val;
            head = head.next;
        }
        rightCol--;

        for(let col = rightCol; col >= leftCol && head; col--){
            matrix[bottomRow][col] = head.val;
            head= head.next;
        }
        bottomRow--;

        for(let row = bottomRow; row >= topRow && head; row--){
            matrix[row][leftCol] = head.val;
            head = head.next;
        }
        leftCol++;
    }
    return matrix;
};
